File: 206-reverse-linked-list/206-reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
    
    ListNode* solve(ListNode* head){
        ListNode* d = NULL;
        
        while(head != NULL){
            ListNode* next = head->next;
            head->next = d;
            d = head;
            head = next;
        }
        
        return d;
    }
    
public:
    ListNode* reverseList(ListNode* head) {
        return solve(head);
    }
};

File: 206-reverse-linked-list/206-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None:
            return head
        if head.next is None:
            return head
        
        dummy, nxt = None, None # Intialising dummy and nxt
        curNode = head # head node ko current Node --> curNode bol diya
        
        # Given linked list say --> 1 -> 2 -> None
        while curNode is not None:
            nxt = curNode.next #pehle asli next node store karli
            # 1     ->   2     ->    None
            # ^curNode   ^ nxt
            curNode.next = dummy 
            # None   <-  1      2    ->    None
            # ^ dummy    ^ curNode
            dummy = curNode
            # None   <-  1      2    ->    None
            #            ^ dummy ab ye hai jo pehle curNode tha
            curNode = nxt
            # None   <-  1      2    ->    None
            #            ^dummy ^ curNode ab ye hai, jo pehle nxt tha
            
        return dummy